### PR TITLE
Remove volume instructions to exclude caching on host

### PR DESCRIPTION
**Pull request by vitalyzhakov**
Hello!

We using docker-compose to debug application.
Defined volume caching on host and not changed after `docker-compose --build` command.

I don't see purpose for defining volume, but removing it will increase debug speed, i think.

### CODE REVIEW

**Comment by vitalyzhakov** [comment]
@LeoColomb , hello!
Can you merge it to master branch?

**Comment by LeoColomb** [comment]
Thanks for the PR @vitalyzhakov.

I need to review your change before merging! 🙂

**Comment by vitalyzhakov** [comment]
@LeoColomb , hello!
How about review timeout?
PR waiting 2 weeks your review.

It consists one string in four files, i think, it does not take much time.

**Comment by LeoColomb** [comment]
> How about review timeout?

Haha, oh, we have a timeout for reviews?  
Don't forget our work is on our free time. 😉

**Comment by vitalyzhakov** [comment]
`That said you can alter content by editing files in the volume directly.`
@LeoColomb , hello!
Do you really think, that users of this image shall change preference by editing volume on host?

I think, the docker-way is make own container based on this container.

**Comment by LeoColomb** [comment]
I'm not sure to follow you, volumes are created for persistence storage and not for inherited container files. Container based on other container should not be the way to store persistant user files. 
Right?

**Comment by vitalyzhakov** [comment]
Persistent storage (such as images, for example) must be external and have backup copy at least.
Users can do this by volume (or mount) instruction when container run https://docs.docker.com/storage/volumes/#choose-the--v-or---mount-flag on any point.
This ability **does not depends on VOLUME** instruction in Dockerfile.
So, users can do that they want :)

On the other hand, VOLUME instruction in Dockerfile prevent debugging with `docker-compose --build `command.
That's why I want to remove it.

**Comment by LeoColomb** [comment]
@J0WI @tianon I need your thoughts on this case, I don't have enough knowledge to go further.

`VOLUME` is here because present on the `Dockerfile` of WordPress official image.  
If the instruction is relevant for WordPress, it is the same for YOURLS, the usage is the exact same.

@vitalyzhakov suggests to remove it, as useless. What are the arguments against (or in favor, things can change :wink:)?

Thanks a lot for your help!

**Comment by J0WI** [comment]
Are the sources mutable? E.g. WordPress can auto update, install plugins and themes etc.
A `VOLUME` might be easier to connect the FPM variants to a webserver container.

**Comment by LeoColomb** [comment]
@J0WI The logic is the exact same from WordPress. Then yes, the sources are mutable.

**Comment by vitalyzhakov** [comment]
@J0WI , @LeoColomb , hello!
Thanks for your time.

Firstly, @LeoColomb , I don`t find a way to install plugin or update yourls in volume.
Is it documented anywhere? I have installed plugins by extending your container as basic and using `COPY` instructions to move plugins into container. Am I right?

Secondly, I want to say that docker helps continuous delivery. I mean, if your want to update soft in the container, use new version on container. It is easy. You don't need volume.

Updating software in volume blocks running application correctly in modern cloud environment (such as kubernetes or docker swarm), because you need to store volume state in external storage (by default volume is stored on node with running container). If node with running container is down, manager runs container on another node with no old volume information. Updating by changing docker image version does not need external volume store and will be more predictable and stable.

Thirdly, I think, external article https://12factor.net/build-release-run will confirm my opinion and we will make better software.

**Comment by J0WI** [comment]
See also https://github.com/docker-library/wordpress/issues/232

**Comment by LeoColomb** [comment]
Ah, thanks a lot @J0WI!

@vitalyzhakov You can find the whole explanation on docker-library/wordpress#232.
Up to now YOURLS follow WordPress' behavior, so Docker usage will be the same.

I'm closing this issue, but feel free to continue the discussion.

**Comment by vitalyzhakov** [comment]
Hello!
@LeoColomb , @J0WI , do you agree, that `VOLUME` instruction brings bugs for users about 3 years?

**Comment by bkief** [comment]
@LeoColomb I agree with @vitalyzhakov's comment that user's can use `docker run -v` to mount the html directory. I don't see any value added by predefining the volume. Some feel it indicated which app directories are considered mutable, but I view that value marginal at best.  By including the volume in the build file you remove the user's choice and also block any further customization of downstream yourls images (e.g. customizing frontends/backends). Perhaps the middle ground here is removing the volume from the build file and adding how to mount the html directory to the documentation.

**Comment by bkief** [comment]
I'll also add that I found it shocking that the WordPress image created a volume so that it can auto-update and persist the new version of WordPress. This cuts against the whole idea of immutable containers. IMHO, this is why `docker pull` and tags exists - you can upgrade the image. Not sure if that is a pattern you want to copy...

**Comment by LeoColomb** [comment]
@vitalyzhakov @bkief Thanks for your inputs. I can understand you point of view, but I would prefer to keep technical guidance from the WordPress repo.

I'm reopening this one to track the activity though.

**Comment by vitalyzhakov** [comment]
@LeoColomb , hello!
There are no activity in wordpress repo to solve this problem.
Who else may approve this PR?

**Comment by LeoColomb** [comment]
I'm afraid only me can review this PR, haha! 😄 
Let us another week to review impact and I'll probably merge this.

File: fpm/Dockerfile [hunk 1]
```diff
@@ -14,9 +14,6 @@ RUN { \
         echo 'opcache.fast_shutdown=1'; \
     } > /usr/local/etc/php/conf.d/opcache-recommended.ini
 
-
```
Comment (LeoColomb) [inline fpm/Dockerfile]: One extra line removed

Comment (vitalyzhakov) [reply fpm/Dockerfile]: Ok, changes are synced.

Comment (LeoColomb) [reply fpm/Dockerfile]: Nice thanks!

**Pull request merged**